locust/test/test_runners.py

```diff
@@ -784,7 +784,56 @@ def test_spawn_fewer_locusts_than_workers(self):
                 num_users += msg.data["num_users"]
             
             self.assertEqual(2, num_users, "Total number of locusts that would have been spawned is not 2")
+
+    def test_custom_shape(self):
+        class MyUser(User):
+            wait_time = constant(0)
+            @task
+            def my_task(self):
+                pass
+    
+        class TestShape(LoadTestShape):
+            def tick(self):
+                run_time = self.get_run_time()
+                if run_time < 2:
+                    return (1, 1, False)
+                elif run_time < 4:
+                    return (2, 2, False)
+                else:
+                    return (0, 0, True)
+    
+        self.environment.user_classes = [MyUser]
+        self.environment.shape_class = TestShape()
+    
+        with mock.patch("locust.rpc.rpc.Server", mocked_rpc()) as server:
+            master = self.get_runner()
+            for i in range(5):
+                server.mocked_send(Message("client_ready", None, "fake_client%i" % i))
+    
+            # Start the shape_worker
+            self.environment.shape_class.reset_time()
+            master.start_shape()
+            sleep(0.5)
+    
+            # Wait for shape_worker to update user_count
+            num_users = 0
+            for _, msg in server.outbox:
+                if msg.data:
+                    num_users += msg.data["num_users"]
+            self.assertEqual(1, num_users, "Total number of users in first stage of shape test is not 1: %i" % num_users)
     
+            # Wait for shape_worker to update user_count again
+            sleep(2)
+            num_users = 0
+            for _, msg in server.outbox:
+                if msg.data:
+                    num_users += msg.data["num_users"]
+            self.assertEqual(3, num_users, "Total number of users in second stage of shape test is not 3: %i" % num_users)
```

Comment:
Maybe also test ramp down?

Reply:
Good idea. I'll add another test. I had a problem that the test only runs for 5 seconds and is then stopped, regardless of what I set. I couldn't see why. Is that correct?

Reply:
OK now there are 2 tests, one for scaling up and one for scaling down. Ideally they would be in the same test but that 5s issue makes it too tight.

Reply:
Hmm... I didnt know there was a 5 second limit and cant really understand where it comes from :)